### PR TITLE
Add Canny edge detection

### PR DESCRIPTION
**Pull request by priba**
### Description
First implementation of the Canny edge detector as mentioned in issue #446. I've not been able to match exactly the results obtained by OpenCV or skimage testing locally even though the results are close. I suspect the issue is likely to be in the non-maximum suppression according to the directions.

I've find this two implementations useful for coding it:
* https://towardsdatascience.com/implement-canny-edge-detection-from-scratch-with-pytorch-a1cccfa58bed 
* https://github.com/DCurro/CannyEdgePytorch/blob/master/net_canny.py

Also, the outputs of the model should be decided as the last step of Threshold & Hysteresis is not differentiable (even though is the one I am testing to get the same results as OpenCV or skimage).

All the test have to be implemented.

### Status
**Work in progress**

### Types of changes

- [] Bug fix (non-breaking change which fixes an issue)
- [x ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

**Comment by edgarriba** [comment]
@ducha-aiki @lferraz can you help with revision ?

**Comment by ducha-aiki** [comment]
Sure, will do tomorrow

File: kornia/filters/canny.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            high: torch.Tensor = magnitude > high_threshold
+
+            edges: torch.Tensor = low * 0.5 + high * 0.5
+
+            if hysteresis:
+
+                edges_old: torch.Tensor = torch.zeros(edges.shape, device=edges.device, dtype=edges.dtype)
+                hysteresis_kernels: torch.Tensor = get_hysteresis_kernel(device, dtype)
+
+                while ((edges_old - edges).abs() != 0).any():
```
Comment (ducha-aiki) [inline kornia/filters/canny.py:108]: Isn`t this while loop slows down everything?
No offence/haven't done tests, just asking

**Comment by edgarriba** [comment]
@priba @ducha-aiki any decision here ?

**Comment by ducha-aiki** [comment]
@edgarriba LGTM, overall

**Review by ducha-aiki — APPROVED** [review]

**Review by edgarriba — COMMENTED** [review]
@priba please fix tests so that we can merge

**Comment by priba** [comment]
I will try to fix it during this week.

File: kornia/filters/canny.py (hunk 1 above)
Comment (priba) [reply kornia/filters/canny.py:108]: Yes, I agree, and it would be nice to avoid it. This is linked to the following issue #866  that should provides a better tool to do this step.

Comment (edgarriba) [reply kornia/filters/canny.py:108]: @lferraz did some tricks with vmap/pmap around this for CLAHE that might be applicable here too. We can leave this a future improvement. Add a TODO note on behalf of that

File: kornia/filters/canny.py [hunk 2]
```diff
@@ -186,22 +214,22 @@ def __init__(
...
+            str(self.sigma) +
+            'low_threshold=' +
+            str(self.low_threshold) +
+            'high_threshold=' +
+            str(self.high_threshold) +
+            'hysteresis=' +
+            str(self.hysteresis) +
+            'eps=' +
+            str(self.eps) +
+            ')'
```
Comment (edgarriba) [inline kornia/filters/canny.py]: @priba try this here https://github.com/kornia/kornia/issues/1053

Comment (priba) [reply kornia/filters/canny.py]: Done. Keep in mind that it will print training=True. from nn.Module

**Review by edgarriba — APPROVED** [review]
